compiler/hash-layout/src/ty.rs

```diff
@@ -0,0 +1,1136 @@
+        ret_ty: ReprTyId,
+        attr_id: ast::AstNodeId,
+    ) -> Self {
+        Self {
+            name,
+            is_intrinsic: false,
+            params,
+            source,
+            ret_ty,
+            polymoprhpic_origin: false,
+            abi: Abi::Hash,
+            attr_id,
+        }
+    }
+
+    /// Get the name from the instance.
+    pub fn name(&self) -> Identifier {
+        self.name
+    }
+
+    /// Check if the instance is of a generic origin.
+    pub fn is_origin_polymorphic(&self) -> bool {
+        self.polymoprhpic_origin
+    }
+
+    /// Check if the [Instance] is an intrinsic function.
+    pub fn is_intrinsic(&self) -> bool {
+        self.is_intrinsic
+    }
+}
+
+static_single_store!(
+    store = pub InstanceStore,
+    id = pub InstanceId,
+    value = Instance,
+    store_name = instances,
+    store_source = repr_stores(),
+    derives = Debug
+);
+
+/// Reference kind, e.g. `&T`, `&mut T`, `&raw T` or `Rc<T>`.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum RefKind {
+    /// Normal reference kind, e.g. `&T` or `&mut T`
+    Normal,
+
+    /// Raw reference kind e.g. `&raw T`
+    Raw,
+
+    /// Reference counted reference kind.
+    Rc,
+}
+
+impl fmt::Display for RefKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RefKind::Normal => write!(f, ""),
+            RefKind::Raw => write!(f, "raw "),
+            RefKind::Rc => write!(f, "rc "),
+        }
+    }
+}
+
+/// Simplified type structure used by the IR and other stages to reason about
+/// Hash programs once types have been erased and simplified.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum ReprTy {
+    /// A primitive signed integer type, e.g. `i32`.
+    Int(SIntTy),
+
+    /// A primitive unsigned integer type, e.g. `u32`.
+    UInt(UIntTy),
+
+    /// A primitive floating point type, e.g. `f32`.
+    Float(FloatTy),
+
+    /// String type.
+    Str,
+
+    /// Boolean type.
+    Bool,
+
+    /// Character type.
+    Char,
+
+    /// The never type
+    Never,
+
+    /// A reference type, referring to a another type, e.g. `&T`, `&mut T`
+    /// or `&raw T`, or `Rc<T>`.
+    Ref(ReprTyId, Mutability, RefKind),
+
+    /// A slice type, `&[T]`.
+    Slice(ReprTyId),
+
+    /// An array type with a specified length, i.e. `[T; N]`
+    Array { ty: ReprTyId, length: usize },
+
+    /// An abstract data structure type, i.e a `struct` or `enum`, or any
+    /// other kind of type.
+    Adt(AdtId),
+
+    /// A function type, with interned parameter type list and a return
+    /// type.
+    Fn {
+        /// The parameter types of the function.
+        params: ReprTyListId,
+
+        /// The return type of the function.
+        return_ty: ReprTyId,
+    },
+
+    /// A function definition, it has an associated instance which denotes
+    /// information about the function, such as the name, defining module,
+    /// ABI, etc.
+    FnDef {
```

Comment:
Seems really strange to me to have this as part of ReprTy.

Reply:
Surely all the `Instance` stuff should be IR-specific?

Reply:
Well yes it ideally should be part of IR, but it still has to be accounted for during layout and other things. It is a special variant specific to IR+ stages

Reply:
But is the layout of a FnDef not the same as Fn? Shouldn't IrTy wrap ReprTy and potentially add some more variants

Reply:
FnDef is effectively a `unit` type because it doesn't really need any storage (mostly because you can't really "assign" a fn-def as an LValue), hence in the layout computation logic it is treated as a `()`.

Reply:
Okay. Do you think in the future it is worth investigating to remove FnDef entirely? It still seems to me that it is not really a "type" in the usual sense, but something else entirely.